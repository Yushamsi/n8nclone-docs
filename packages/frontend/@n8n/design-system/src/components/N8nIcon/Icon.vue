<script lang="ts" setup>
import { computed, useCssModule } from 'vue';

import type { IconSize, IconColor } from '@n8n/design-system/types/icon';

import type { IconName } from './icons';
import { allIcons } from './icons';

interface IconProps {
	icon: IconName;
	size?: IconSize | number;
	spin?: boolean;
	color?: IconColor;
}

defineOptions({ name: 'N8nIcon' });

const props = withDefaults(defineProps<IconProps>(), {
	spin: false,
	size: undefined,
	color: undefined,
});

const $style = useCssModule();
const classes = computed(() => {
	const applied: string[] = [];
	if (props.spin) {
		applied.push('spin');
	}

	return applied.map((c) => $style[c]);
});

const sizesInPixels: Record<IconSize, number> = {
	xsmall: 10,
	small: 12,
	medium: 14,
	large: 16,
	xlarge: 20,
};

const size = computed((): { height: string; width: string } => {
	let sizeToApply = '1em';
	if (props.size) {
		sizeToApply = `${typeof props.size === 'number' ? props.size : sizesInPixels[props.size]}px`;
	}

	return {
		height: sizeToApply,
		width: sizeToApply,
	};
});

const styles = computed(() => {
	const stylesToApply: Record<string, string> = {};

	if (props.color) {
		stylesToApply.color = `var(--color-${props.color})`;
	}

	return stylesToApply;
});
</script>

<template>
	<Component
		:is="allIcons[icon as keyof typeof allIcons]"
		v-if="allIcons[icon as keyof typeof allIcons]"
		:class="classes"
		aria-hidden="true"
		focusable="false"
		role="img"
		:height="size.height"
		:width="size.width"
		:data-icon="props.icon"
		:style="styles"
	/>
</template>

<style>
.n8n-icon {
	display: inline-flex;
	justify-content: center;
	align-items: center;
}
</style>

<style lang="scss" module>
.spin {
	animation: spin 1s linear infinite;
}

.primary {
	color: var(--color-primary);
}

@keyframes spin {
	from {
		transform: rotate(0deg);
	}
	to {
		transform: rotate(360deg);
	}
}
</style>
