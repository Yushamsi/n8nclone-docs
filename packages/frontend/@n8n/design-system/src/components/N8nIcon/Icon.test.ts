import { render } from '@testing-library/vue';

import Icon from './Icon.vue';

describe('Icon', () => {
	it('should render correctly with default props', () => {
		const wrapper = render(Icon, {
			props: {
				icon: 'check',
			},
		});
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('should render correctly with a custom size', () => {
		const wrapper = render(Icon, {
			props: {
				icon: 'check',
				size: 24,
			},
		});
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('should render correctly with predefined size', () => {
		const wrapper = render(Icon, {
			props: {
				icon: 'check',
				size: 'large',
			},
		});
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('should render correctly with spin enabled', () => {
		const wrapper = render(Icon, {
			props: {
				icon: 'check',
				spin: true,
			},
		});
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('should render correctly with a custom color', () => {
		const wrapper = render(Icon, {
			props: {
				icon: 'check',
				color: 'primary',
			},
		});
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('should render correctly with all props combined', () => {
		const wrapper = render(Icon, {
			props: {
				icon: 'check',
				size: 'medium',
				spin: true,
				color: 'secondary',
			},
		});
		expect(wrapper.html()).toMatchSnapshot();
	});
});
