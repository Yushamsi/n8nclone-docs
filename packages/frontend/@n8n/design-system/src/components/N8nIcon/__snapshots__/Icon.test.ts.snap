// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Icon > should render correctly with a custom color 1`] = `
"<svg viewBox="0 0 24 24" width="1em" height="1em" class="" aria-hidden="true" focusable="false" role="img" data-icon="check">
  <path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M20 6L9 17l-5-5"></path>
</svg>"
`;

exports[`Icon > should render correctly with a custom size 1`] = `
"<svg viewBox="0 0 24 24" width="24px" height="24px" class="" aria-hidden="true" focusable="false" role="img" data-icon="check">
  <path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M20 6L9 17l-5-5"></path>
</svg>"
`;

exports[`Icon > should render correctly with all props combined 1`] = `
"<svg viewBox="0 0 24 24" width="14px" height="14px" class="spin" aria-hidden="true" focusable="false" role="img" data-icon="check">
  <path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M20 6L9 17l-5-5"></path>
</svg>"
`;

exports[`Icon > should render correctly with default props 1`] = `
"<svg viewBox="0 0 24 24" width="1em" height="1em" class="" aria-hidden="true" focusable="false" role="img" data-icon="check">
  <path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M20 6L9 17l-5-5"></path>
</svg>"
`;

exports[`Icon > should render correctly with predefined size 1`] = `
"<svg viewBox="0 0 24 24" width="16px" height="16px" class="" aria-hidden="true" focusable="false" role="img" data-icon="check">
  <path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M20 6L9 17l-5-5"></path>
</svg>"
`;

exports[`Icon > should render correctly with spin enabled 1`] = `
"<svg viewBox="0 0 24 24" width="1em" height="1em" class="spin" aria-hidden="true" focusable="false" role="img" data-icon="check">
  <path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M20 6L9 17l-5-5"></path>
</svg>"
`;
